### PR TITLE
Restructured BEB reconciler

### PR DESCRIPTION
**Pull request by mfaizanse**
**Description**

Changes proposed in this pull request:

- Restructured the BEB reconciler for better code quality and to reduce number of reconciliation calls.

**Related issue(s)**

### CODE REVIEW

File: components/eventing-controller/api/v1alpha1/subscription_lifecycle.go [hunk 1]
```diff
@@ -76,6 +76,24 @@ func (s SubscriptionStatus) IsReady() bool {
...
+		if conditionType == condition.Type {
+			return &condition
+		}
+	}
+	return nil
+}
+
+// ShouldUpdateReadyStatus checks if there is a mismatch between the
+// subscription Ready Status and the Ready status of all the conditions
+func (s SubscriptionStatus) ShouldUpdateReadyStatus() bool {
```
Comment (pxsalehi) [inline components/eventing-controller/api/v1alpha1/subscription_lifecycle.go:90]: Should we have some unit tests for these two new methods? At least for `ShouldUpdateReadyStatus`?

File: components/eventing-controller/controllers/subscription/beb/equality.go [hunk 2]
```diff
@@ -1,6 +1,8 @@
 package beb
```
Comment (pxsalehi) [inline components/eventing-controller/controllers/subscription/beb/equality.go:1]: How about a `components/eventing-controller/controllers/subscription/beb/equality_test.go`?  :-)

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 3]
```diff
@@ -869,29 +884,30 @@ func isInDeletion(subscription *eventingv1alpha1.Subscription) bool {
 const timeoutRetryActiveEmsStatus = time.Second * 30
```
Comment (pxsalehi) [inline components/eventing-controller/controllers/subscription/beb/reconciler.go:884]: What do you think about moving all the consts up in the file in one block?

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 4]
```diff
@@ -98,199 +99,229 @@ func NewReconciler(ctx context.Context, client client.Client, applicationLister
...
+		if updateErr := r.updateSubscription(ctx, subscription, log); updateErr != nil {
+			return ctrl.Result{}, errors.Wrap(err, updateErr.Error())
 		}
+		return ctrl.Result{}, errors.Wrap(err, "sync finalizer failed")
 	}
 
-	// mark if the subscription status was changed
-	statusChanged := false
+	// sync APIRule for the desired subscription
+	apiRule, err := r.syncAPIRule(ctx, subscription, log)
```
Comment (pxsalehi) [inline components/eventing-controller/controllers/subscription/beb/reconciler.go:136]: One question: it seems in the code on main, the API rule logics happen earlier in the reconcile call. In your PR, they have been move down. Is that correct? If yes, why?

**Review by pxsalehi — DISMISSED** [review]
Thanks for cleaning up the BEB reconciler!

File: components/eventing-controller/controllers/subscription/beb/reconciler.go (hunk 4 above)
Comment (mfaizanse) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go:137]: No, We have followed the same sequence as in main:
1. syncFinalizer
2. syncAPIRule
3. syncBEBSubscription

File: components/eventing-controller/controllers/subscription/beb/reconciler.go (hunk 3 above)
Comment (mfaizanse) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go]: moved

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 5]
```diff
@@ -98,199 +99,229 @@ func NewReconciler(ctx context.Context, client client.Client, applicationLister
 // +kubebuilder:rbac:groups=gateway.kyma-project.io,resources=apirules,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:printcolumn:name="Ready",type=bool,JSONPath=`.status.Ready`
 
-// TODO: Optimize number of reconciliation calls in eventing-controller #9766: https://github.com/kyma-project/kyma/issues/9766
 func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (VladislavPaskar) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go:104]: It is already done in [this](https://github.com/kyma-project/kyma/pull/13315/files#diff-73dbd6913962be0ebf2cc929722f3d75590b312285538a06e981937bdb213521L49) PR, which is already in review

File: components/eventing-controller/api/v1alpha1/subscription_lifecycle.go (hunk 1 above)
Comment (raypinto) [reply components/eventing-controller/api/v1alpha1/subscription_lifecycle.go:90]: done

File: components/eventing-controller/controllers/subscription/beb/equality.go (hunk 2 above)
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/beb/equality.go:1]: Done :)

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 6]
```diff
@@ -641,77 +684,57 @@ func getSvcNsAndName(url string) (string, string, error) {
...
+	for _, condition := range expectedConditionsMap {
+		requiredConditions = append(requiredConditions, condition)
 	}
 
-	r.emitConditionEvent(subscription, condition)
-	return nil
+	return requiredConditions
 }
 
 // replaceStatusCondition replaces the given condition on the subscription. Also it sets the readiness in the status.
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go:742]: I think with [these tests](https://github.com/kyma-project/kyma/pull/13314#discussion_r801765738) added to the `unit_test.go` file, the file is going to grow and I think it should stay.

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 7]
```diff
@@ -98,199 +99,229 @@ func NewReconciler(ctx context.Context, client client.Client, applicationLister
...
+}
 
-	statusChangedAtCheck, retry, errTimeout := r.checkStatusActive(subscription)
-	statusChanged = statusChanged || statusChangedAtCheck
-	if errTimeout != nil {
-		logger.Errorw("timeout at retry", "error", errTimeout)
-		result.Requeue = false
-		return statusChanged, errTimeout
+// syncConditionSubscribed syncs the condition ConditionSubscribed
+func (r *Reconciler) syncConditionSubscribed(subscription *eventingv1alpha1.Subscription, isSubscribed bool) {
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go:298]: Agree!! Done as far as possible :)

File: components/eventing-controller/controllers/subscription/beb/reconciler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pxsalehi) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go:733]: My point was not to remove the file, but to give it a more meaningful name. I think per Go conventions, `unit_test.go` is supposed to contain the tests for a file called `unit.go`, which doesn't exist. But it doesn't matter, we can keep it like that.

**Review by pxsalehi — APPROVED** [review]

**Pull request merged**